crates/switch-mm/src/shift.rs

```diff
@@ -0,0 +1,221 @@
+                        Default::default(),
+                    ));
+                    instrs.push((
+                        Instr::Const(Const {
+                            value: Value::I32(OFFSET),
+                        }),
+                        Default::default(),
+                    ));
+                    instrs.push((
+                        Instr::Binop(Binop {
+                            op: BinaryOp::I32Add,
+                        }),
+                        Default::default(),
+                    ));
+                    instrs.push((
+                        Instr::LocalGet(LocalGet { local: local1 }),
+                        Default::default(),
+                    ));
+                    instrs.push((instr, loc));
+                }
+                Instr::Load(Load {
+                    memory,
+                    kind,
+                    arg: MemArg { align, offset },
+                }) => {
+                    let offset = offset.checked_add(OFFSET as u32).unwrap();
+                    let instr = Instr::Load(Load {
+                        memory,
+                        kind,
+                        arg: MemArg { align, offset },
+                    });
+                    instrs.push((instr, loc));
+                }
+                Instr::Store(Store {
+                    memory,
+                    kind,
+                    arg: MemArg { align, offset },
+                }) => {
+                    let offset = offset.checked_add(OFFSET as u32).unwrap();
+                    let instr = Instr::Store(Store {
+                        memory,
+                        kind,
+                        arg: MemArg { align, offset },
+                    });
+                    instrs.push((instr, loc));
+                }
+                Instr::AtomicFence(_) | Instr::AtomicNotify(_) | Instr::AtomicWait(_) | Instr::AtomicRmw(_) | Instr::Cmpxchg(_) => todo!(),
+                Instr::LoadSimd(_) => todo!(),
+                _ => instrs.push((instr, loc)),
+            }
+        }
+        seq.instrs = instrs;
+    }
+}
+fn get_local(gen: &mut ModuleLocals, locals: &mut Vec<LocalId>, idx: usize) -> LocalId {
+    if idx < locals.len() {
+        locals[idx]
+    } else {
+        for _ in locals.len()..=idx {
+            locals.push(gen.add(ValType::I32));
+        }
+        locals[idx]
+    }
+}
+pub fn shift_adapter_module(module: &mut Module) -> Result<()> {
+    let stack_pointer = module
+        .globals
+        .iter()
+        .find_map(|g| {
+            let name = g.name.clone()?;
+            if name == "__stack_pointer" {
+                Some(g.id())
+            } else {
+                None
+            }
+        })
+        .unwrap();
+    module.globals.get_mut(stack_pointer).kind =
+        GlobalKind::Local(ConstExpr::Value(Value::I32(64 * 1024)));
+    Ok(())
+}
+pub fn shift_main_module(module: &mut Module) -> Result<()> {
```

Comment:
Were you aware of the module-rewriting `Reencode` trait, which exposes hooks for instructions, data sections, etc.? I used it [for removing duplicate imports ](https://github.com/bytecodealliance/wasm-tools/commit/d2c36f43e88317de245fb128540bec6303ee0748#diff-6abaa3c2edecec8fe38805ba402e7400e31ed3e1f3d766eebbedec40ada5e110R143) and found it made the code very easy to follow. I don't know whether it would improve your code here or not, but I thought I'd mention it.